pkg/planner/memo/task_scheduler.go

```diff
@@ -0,0 +1,53 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memo
+
+var _ TaskScheduler = &SimpleTaskScheduler{}
+
+// TaskScheduler is a scheduling interface defined for serializing(single thread)/concurrent(multi thread) running.
+type TaskScheduler interface {
+	ExecuteTasks()
+}
+
+// SimpleTaskScheduler is defined for serializing scheduling of memo tasks.
+type SimpleTaskScheduler struct {
+	Err          error
+	SchedulerCtx TaskSchedulerContext
+}
+
+// ExecuteTasks implements the interface of TaskScheduler.
+func (s *SimpleTaskScheduler) ExecuteTasks() {
+	stack := s.SchedulerCtx.getStack()
+	defer func() {
+		// when step output of the scheduler, if the stack is empty, clean and release it.
+		if !stack.Empty() {
+			stack.Destroy()
+		}
+	}()
+	for !stack.Empty() {
+		// when use customized stack to drive the tasks, the call-chain state is dived in the stack.
+		task := stack.Pop()
+		if err := task.execute(); err != nil {
+			s.Err = err
+			return
+		}
+	}
+}
+
+// TaskSchedulerContext is defined for scheduling logic calling, also facilitate interface-oriented coding and testing.
+type TaskSchedulerContext interface {
+	getStack() *TaskStack
+	pushTask(task Task)
```

Comment:
We can change the array from `[]task` to `[]*task` to avoid the overhead of pass by value.
Maybe we can make a benchmark later for the two kinds of usage.

If the task object is not small, it would be better to allocate it once on the heap instead of passing by value and allocating it at stack when we enter the `getTask` and `pushTask`.

Reply:
Task is a interface definition
```
type Task interface {
	// task self executing logic
	execute() error
	// task self description string.
	desc() string
}
type TaskStack struct {
	tasks []Task
}
```

Reply:
https://github.com/pingcap/tidb/pull/51866/commits/8b62caef209f740c505fb02025444d6763299b96

Reply:
// BenchmarkTestStack2Pointer-8   	   43802	     27569 ns/op	   24000 B/op	    2000 allocs/op
// BenchmarkTestStackInterface-8   	  108644	     10736 ns/op	    8000 B/op	    1000 allocs/op